packages/shared/src/server/repositories/clickhouse.ts

```diff
@@ -1,34 +1,106 @@
 import { env } from "../../env";
-import { clickhouseClient } from "../clickhouse/client";
+import {
+  clickhouseClient,
+  convertDateToClickhouseDateTime,
+} from "../clickhouse/client";
 import { logger } from "../logger";
-import { getCurrentSpan, instrumentAsync } from "../instrumentation";
+import { instrumentAsync } from "../instrumentation";
+import { S3StorageService } from "../services/S3StorageService";
+import { randomUUID } from "crypto";
+import { getClickhouseEntityType } from "../clickhouse/schemaUtils";
 
-export const convertRecordToJsonSchema = (
-  record: Record<string, string>,
-): JsonNested | undefined => {
-  const jsonSchema: JsonNested = {};
+let s3StorageServiceClient: S3StorageService;
 
-  // if record is empty, return undefined
-  if (Object.keys(record).length === 0) {
-    return undefined;
+const getS3StorageServiceClient = (bucketName: string): S3StorageService => {
+  if (!s3StorageServiceClient) {
+    s3StorageServiceClient = new S3StorageService({
+      bucketName,
+      accessKeyId: env.LANGFUSE_S3_EVENT_UPLOAD_ACCESS_KEY_ID,
+      secretAccessKey: env.LANGFUSE_S3_EVENT_UPLOAD_SECRET_ACCESS_KEY,
+      endpoint: env.LANGFUSE_S3_EVENT_UPLOAD_ENDPOINT,
+      region: env.LANGFUSE_S3_EVENT_UPLOAD_REGION,
+      forcePathStyle: env.LANGFUSE_S3_EVENT_UPLOAD_FORCE_PATH_STYLE === "true",
+    });
   }
+  return s3StorageServiceClient;
+};
+
+export async function upsertClickhouse(opts: {
+  table: "scores"; // TODO: Modify eventType logic to support more tables going forward
+  records: Record<string, unknown>[];
+}): Promise<void> {
+  return await instrumentAsync({ name: "clickhouse-upsert" }, async (span) => {
+    // https://opentelemetry.io/docs/specs/semconv/database/database-spans/
+    span.setAttribute("ch.query.table", opts.table);
+
+    // drop trailing s and pretend it's always a create.
+    // Only applicable to scores (and eventually traces).
+    const eventType = `${opts.table.slice(0, -1)}-create`;
 
-  for (const key in record) {
-    try {
-      jsonSchema[key] = JSON.parse(record[key]);
-    } catch (e) {
-      jsonSchema[key] = record[key];
+    // If event upload is enabled, we store all rows in S3 to have a backup
+    if (env.LANGFUSE_S3_EVENT_UPLOAD_ENABLED === "true") {
+      if (env.LANGFUSE_S3_EVENT_UPLOAD_BUCKET === undefined) {
+        throw new Error("S3 event store is enabled but no bucket is set");
+      }
+      const s3Client = getS3StorageServiceClient(
+        env.LANGFUSE_S3_EVENT_UPLOAD_BUCKET,
+      );
+      await Promise.all(
+        opts.records.map((record) => {
+          s3Client.uploadJson(
+            `${env.LANGFUSE_S3_EVENT_UPLOAD_PREFIX}${record.project_id}/${getClickhouseEntityType(eventType)}/${record.id}/${randomUUID()}.json`,
+            [
+              {
+                id: randomUUID(),
+                timestamp: new Date().toISOString(),
+                type: eventType,
+                body: record,
```

Comment:
@maxdeichmann The body we write to S3 here has the Clickhouse format (i.e. using snake_case). Other events follow the zod schema. I could try to use our generic mapping functions to move back and forth between them here, but this will add more table specific logic to this generic upsert. Do you have a gut feel whether we should handle it in here or do some more explicit mapping on the respective repositories?